### ISSUE TITLE
[Feature] Transactions list screen

### ISSUE THREAD

**Issue by talmahaj** [tari-project/tari#1003]
**User Stories**

> Each time I want to know how much Tari I currently have available to spend, I want it to be effortless to see my current balance

> When I want to refer back to any particular transaction to check its status or amount -- or see how it may be influencing my current balance -- I need to be able to see all my prior transactions as well as the attributes by which I'd recognize them (e.g. date, contact, amount, note content)

-----

**UX/UI**

1. Transaction list expansion animations: Download the Principle app and open this file in it from your iPhone (unfortunately there is no Android support, but there is a 14-day free trial of the Mac app. @Karim-tari will record a little video next time):
https://www.dropbox.com/s/xl4s77sl1mgaua3/Transaction%20Flow.prd?dl=0
**Dec 1 Update: Note that the expanded list UI here is outdated, the full screen version should have a sticky header and a close button on the top left per the Zeplin links below**

2. Pls use the following Zeplin links for the screen specs in 3 different states of the interaction in the prototype -- they're in 1x for iOS so just convert 1:1 from pt to dp for Android.
https://zpl.io/bLnx4OQ
https://zpl.io/aNn9wmv
https://zpl.io/V4dv0OX

Assets should be in Dropbox folder, pls let Karim know if any are missing.

-----

**Open Questions**

All answered in comments and in spec now.

~~1. @Karim-tari How should the animation work for pending transactions?~~

~~2. @Karim-tari The list items are a bit hard to read in the prototype; can you also create a second version for dev to try out where we bump up the list item text sizes, maybe to the same size as the Recipient screen list items?~~

~~3. @Karim-tari What is the animation as the screen comes into view? I know we had this in the video you made with the prior design, is that still up to date?~~

~~4. @Karim-tari What are the animations when...
    1. the balance changes while the user is already viewing the screen?
    2. a new transaction appears in the list, or a transaction status changes from Pending to complete?~~

~~5. Jay, Kutsal, @philipr-za Not sure how the API pagination works right now but can we load at least 20 list items at a time?~~

-----

**Acceptance Criteria**

Design

1. Final UI styles appear per Dec 3 Updates section: 
https://app.zeplin.io/project/5dbdf1a40bdfd3832e35c9c5/dashboard?seid=5de61c380c863977d92a7d0a

General

1. Page load animations
Whenever the transaction list loads, its elements should still animate in per 00:12/13 in the first prototype we did: https://vimeo.com/369345180 

Top section

1. Available Balance should display only Tari that the user is currently able to spend. That is, exclude the amount of any pending transactions (whether user is sender or receiver).
    1. For the purposes of this ticket, a pending transaction will be defined as any transaction that has been initiated but not yet mined, whether or not it has been received by the recipient or broadcast by the sender yet

2. Dynamic balance changes
Whenever the available balance changes, any changed digits should still animate in per 01:25 in the first prototype we did: https://vimeo.com/369345180**

3. QR code button will not yet be functional; out of scope for this ticket

List

1. The expanded list UI should have a sticky header and a close button on the top left per https://zpl.io/aNn9wmv and https://zpl.io/V4dv0OX

2. As soon as user starts scrolling back up in the tx list, no matter how far down they are, the Send Tari button should immediately appear again.

3. Transaction sections should be named as follows, and only appear if there is 1+ relevant transaction to show in that section according to User's current local time:
    1. "Pending Transactions"
    2. "Today"
    3. "Yesterday"
    4. Weekday, e.g. "Wednesday", for the 5 days preceding "Yesterday"
    5. Mon DD, e.g. "Nov 13", for any transaction preceding the above

4. Transactions within each section should be ordered by the timestamp the transaction was first initiated (descending)

5. When Contact name not available, Emoji ID should display instead with the first 3 and last 3 emojis in the sequence. (Note that there is a gradient over the emojis in the UI so that they fade out and then back in again.)

6. Long notes should be truncated with ellipses

7. Pulling down on list should at least fake a manual refresh

8. Edge case handling for large transactions (…not exactly a tweak, we just didn’t show these cases in prior mocks): https://zpl.io/2yEzvGK
    - Green pill length should be dynamic based on number of digits in amount
    - Truncation point where contact name and note display ellipses (…) should be dynamic based on number of digits in amount
    - Comma should appear when transaction in thousands of Tari
    - Two decimal places should always appear

9. Transaction sections and transactions appear/disappear as follows:
https://vimeo.com/showcase/6606770/video/376982133 pw: t@ri (ignore inconsistent typography, final styles are in Zeplin).   In sequence per the video:
    1. For every new section that appears — whether it is Pending Transactions OR a calendar day — the section heading should animate in at the same time as the transaction(s) within it
    2. When the Pending Transactions section appears, the animation next to the section header on the right should appear at the same time, and persist until there are no more pending transactions (and thus the entire section disappears out of view)
    3. If a section already exists, but it has a new list item, the section heading should first slide to accommodate that item before the new list item animates in
    4. When there are no more pending transactions, progress animation should become green checkmark before section disappears from view
    5. When Pending Transactions section disappears, now-complete transaction should animate into the calendar day section at the same time the whole list slides up

10. Transactions move to the correct sections if I switch my phone’s time zone such that a transaction is now on a prior or future day

-----

**Implementation Notes**

1. Don't worry about the list's empty state for now -- we'll deal with that in a separate ticket once we determine the ideal onboarding flow.

2. Separate tickets will be created for handling the following cases:
- Recipient offline
- More granular pending states

3. Until we design our edge case handling, for now please map all the BE states as follows, noting we may add additional FE states in the next few weeks:
    - Pending:
        - Send to Receiver -> Pending Outbound Transaction
        - Receiver has replied -> Completed transaction (Not broadcast to mempool)
        - Broadcast to Mempool -> Completed transaction (Broadcast)
    - Complete:
        - Mined -> Completed Transaction (Mined)

**Comment by talmahaj** [comment]
Also @Karim-tari can you confirm that the red/green pill on the right side of the list item should collapse and expand as the amount changes?

**Comment by Jasonvdb** [comment]
> Available Balance should display only Tari that the user is currently able to spend. That is, exclude the amount of any pending transactions (whether user is sender or receiver).

@talmahaj so available balance in this scenario won't work that way for mimble wimble reasons. If a user creates a new transaction to send, no matter what the amount will be, the whole available balance is zero, but the pending incoming balance and pending outgoing balance will have values. As it works now, if there are any pending transactions then none of the user's balance is spendable until the transaction is mined.

Not sure if that means we need to reconsider displaying the words "Available Balance"?

**Comment by Jasonvdb** [comment]
> Pulling down on screen should manually refresh it, but the list should already be listening for change notifications and updating dynamically as they come in

Adding pull to refresh is quick to add in but in this case it won't actually be doing anything as the UI list will stay in sync with the wallet lib's list of transactions. The wallet at the moment doesn't have a "refresh" function as its listening for incoming messages. A pull to refresh is helpful for updating data retrieved from a server. 
@philipr-za what do you think? Would there ever be a refresh ffi function?

**Comment by philipr-za** [comment]
The wallet will not be "listening for changes" per se as the base nodes will not be pushing messages to the wallets. The wallets will need to periodically poll the base nodes for updates. So I think a pull to refresh will make sense as it will force a polling cycle now rather than at the background frequency.

**Comment by Jasonvdb** [comment]
Got it thanks, then I'll add pull to refresh in now with a stub function for later when that's ready 👍

**Comment by Karim-tari** [comment]
I was thinking it would always have 2 decimal points. So it will be never be bigger than 50.00 To see the exact amount you have to tap into the transaction detail. 

That being said, I think this is going to depend on the value of Tari. If rounding to 2 decimal points on the feed will be sufficient or not.

Should we have a call to chat through this?

**Comment by talmahaj** [comment]
> I was thinking it would always have 2 decimal points. So it will be never be bigger than 50.00 To see the exact amount you have to tap into the transaction detail. That being said, I think this is going to depend on the value of Tari. If rounding to 2 decimal points on the feed will be sufficient or not. Should we have a call to chat through this?
> […](#)
> On Nov 26, 2019, at 3:54 AM, Tal Flanchraych ***@***.***> wrote: ﻿ Also @Karim-tari can you confirm that the red/green pill on the right side of the list item should collapse and expand as the amount changes? — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or unsubscribe.

@Karim-tari I'm not referring to decimal points -- no issues there -- but to tens vs. hundreds vs. thousands of Tari

**Comment by talmahaj** [comment]
> The wallet will not be "listening for changes" per se as the base nodes will not be pushing messages to the wallets. The wallets will need to periodically poll the base nodes for updates. So I think a pull to refresh will make sense as it will force a polling cycle now rather than at the background frequency.

Got it, thanks @philipr-za! What will be the bg frequency? And how will this change once we get notifications working given that they'll be pushing updates?

**Comment by talmahaj** [comment]
> Got it thanks, then I'll add pull to refresh in now with a stub function for later when that's ready 👍

@Jasonvdb Sounds good -- I was thinking that even if it were never to be functional long-term (e.g. if there were real-time updates), users still like to compulsively refresh in order to give themselves peace of mind, so we could just fake it forever

**Comment by philipr-za** [comment]
> > The wallet will not be "listening for changes" per se as the base nodes will not be pushing messages to the wallets. The wallets will need to periodically poll the base nodes for updates. So I think a pull to refresh will make sense as it will force a polling cycle now rather than at the background frequency.
> 
> Got it, thanks @philipr-za! What will be the bg frequency? And how will this change once we get notifications working given that they'll be pushing updates?

Haven't made a call on that yet, will need to consider how much traffic we are willing for the app to generate but my first instinct is somewhere in the 30-60 second range. 

If we get push notifications working then probably even slower or not at all if the notifications are super reliable.

**Comment by talmahaj** [comment]
> > Available Balance should display only Tari that the user is currently able to spend. That is, exclude the amount of any pending transactions (whether user is sender or receiver).
> 
> @talmahaj so available balance in this scenario won't work that way for mimble wimble reasons. If a user creates a new transaction to send, no matter what the amount will be, the whole available balance is zero, but the pending incoming balance and pending outgoing balance will have values. As it works now, if there are any pending transactions then none of the user's balance is spendable until the transaction is mined.
> 
> Not sure if that means we need to reconsider displaying the words "Available Balance"?

@Jasonvdb @philipr-za Iiiinteresting. Some followup Qs so we can better understand, in case we want to reconsider the design beyond just the copy:

- Can you confirm the definition of pending incoming balance and pending outgoing balance?
- So does this mean a user can only purchase one thing with Tari at a time? And if it happens to take 30-60 minutes to mine, I can't buy anything else until that's over?
- What if a user is a frequent recipient of transactions and there is usually one being mined at a given time? Are they never able to spend their Tari?
- Can you think of any non-MW analogs for this that may be familiar to end users? Trying to figure out how to make this match their existing mental models in an elegant way...

cc @Karim-tari

**Comment by CjS77** [comment]
> > Available Balance should display only Tari that the user is currently able to spend. That is, exclude the amount of any pending transactions (whether user is sender or receiver).
> 
> @talmahaj so available balance in this scenario won't work that way for mimble wimble reasons. If a user creates a new transaction to send, no matter what the amount will be, the whole available balance is zero, but the pending incoming balance and pending outgoing balance will have values. As it works now, if there are any pending transactions then none of the user's balance is spendable until the transaction is mined.
> 
> Not sure if that means we need to reconsider displaying the words "Available Balance"?

It's a little more nuanced. You can think of your Tari wallet as having different sized notes in them. Let's say your balance is T150; made up by a T100 note, a T49 and a T1 note (notes can be of any denomination because math).

When you send someone Tari, the wallet will select a note(s) to spend. When a note is involved in a tx, the _whole_ note is locked up. So this is what @Jasonvdb was referring to. Thus if you're sending T2 and the wallet selects the T100 note to spend, your spendable balance will temporarily drop to T50. 

The T100 note will then be split into a T2 note, a T0.001 note (approx, for fees), and a T97.998 note (your change). When do these new notes become spendable again, you ask?

T0.001 - This goes to a miner who can only spend it after a day or so.

T97.998 - As soon as the sender broadcasts the final transaction to the network, she can technically spend the change. If the receiver is online, it's a few ms between hitting the send button and having the change output available.

T2 - It depends on how much risk you are willing to take. The recipient should wait until it's mined AND wait for a certain number of confirmations before considering the transaction final. But technically, he can spend it as soon as it hits the mempool (which is usually in a few ms if both sender and receiver are online).

**Comment by CjS77** [comment]
> When Contact name not available, Emoji ID should display instead with the first 3 and last 3 emojis in the sequence. (Note that there is a gradient over the emojis in the UI so that they fade out and then back in again.)

Until the emoji API is ready, just display the hex output instead. Emoji and hex are both UTF-8 strings and so should be written as a drop-in replacement. The emoji API will not be done inside the next 2 weeks.

**Comment by philipr-za** [comment]
If we start from a wallet with no pending balance it will have in it a number of UTXO's (outputs) in it that are available to spend. For example let's say it has 10 outputs in it with the following values is ascending order: [10, 20, 50, 55, 100, 150, 250, 300, 500, 1200].

At this stage the balance would be:
- Available Balance: 2635
- Pending Incoming Balance: 0
- Pending Outgoing Balance: 0

Now if this wallet wants to send a transaction of 400 it will need to spend the following outputs for this transaction (like selecting fiat bills from a wallet)

[10, 20, 50, 55, 100, 150, 250] = 635 so 400 will be sent and 235 will be change (assuming zero fees)

At this stage the Sender balance would be:
- Available Balance: 2000
- Pending Incoming Balance: 235
- Pending Outgoing Balance: 400

Assuming the Receiver is a first time receiver with a blank balance after they receive this transaction their balance will be:
- Available Balance: 0
- Pending Incoming Balance: 400
- Pending Outgoing Balance: 0

This transaction is sent to the Receiver who will sign their side of it and send it back, the Sender will complete the transaction and transmit it. The balances will remain pending until the transaction is mined. The respective balances and UTXO sets will then be:

Sender Balance:
 - Available Balance: 2235
 - Pending Incoming Balance: 0
 - Pending Outgoing Balance: 0
 UTXOs: [235, 300, 500, 1200]

 Receiver Balance:
 - Available Balance: 400
 - Pending Incoming Balance: 0
 - Pending Outgoing Balance: 0
 UTXOs: [400]
 
 I think a non MW analog to this process is to consider the UTXOs as fiat bills in a wallet. When you pay for something you have to hand over a set of bills that adds up to more than the value being spent and you will receive change.
 The pending status is a common UX feature of internet banking using a credit card or EFT (at least in South Africa?).

**Comment by talmahaj** [comment]
[PatternAssets.zip](https://github.com/tari-project/tari/files/3896003/PatternAssets.zip)

@Jasonvdb Kutsal the ones with suffix a are android, suffix i are iOS -- hope these work for you.

**Comment by CjS77** [comment]
> Jay, Kutsal, @philipr-za Not sure how the API pagination works right now but can we load at least 20 list items at a time?

Since we're using SQLite on the backend, pagination should be a relatively easy implementation, and so may as well go in from day 1

**Comment by philipr-za** [comment]
> > Jay, Kutsal, @philipr-za Not sure how the API pagination works right now but can we load at least 20 list items at a time?
> 
> Since we're using SQLite on the backend, pagination should be a relatively easy implementation, and so may as well go in from day 1

I will make an issue for this, currently no pagination but easy enough to add.

**Comment by CjS77** [comment]
>I was thinking it would always have 2 decimal points. So it will be never be bigger than 50.00 To see the exact amount you have to tap into the transaction detail.
That being said, I think this is going to depend on the value of Tari. If rounding to 2 decimal points on the feed will be sufficient or not.

I'm ok with this. As long as it's possible to toggle and view the full precision somehow.

**Comment by Jasonvdb** [comment]
@philipr-za would pagination be necessary from your side? I get the `length` of completed transactions and then I'm only fetching the transactions by position when they need to be displayed. Is there a benefit to retrieving batches of transactions instead of loading each one as it's needed if it's all local on the device?

**Comment by philipr-za** [comment]
> @philipr-za would pagination be necessary from your side? I get the `length` of completed transactions and then I'm only fetching the transactions by position when they need to be displayed. Is there a benefit to retrieving batches of transactions instead of loading each one as it's needed if it's all local on the device?

I guess the benefit is how much of the list you copy into memory at any one time but the data being loaded is not huge, I cannot imagine a typical wallet having more than hundreds at most?

@CjS77 What do you think?

**Comment by Jasonvdb** [comment]
So on iOS, and suspect android is similar, if you have a transaction list of 100 elements but only 10 elements are displayed in the scroll view at any time, then only the like 11 rows at most are loaded into memory at once. The rest get loaded in as they're about to scroll into view. So as long as I have the length, I'm only ever going to load 11 (cloned) completed transactions into memory in one go. I don't know if the wallet lib loads everything in at once but on the app side i think this is solved by default 🤷‍♂

**Comment by philipr-za** [comment]
> So on iOS, and suspect android is similar, if you have a transaction list of 100 elements but only 10 elements are displayed in the scroll view at any time, then only the like 11 rows at most are loaded into memory at once. The rest get loaded in as they're about to scroll into view. So as long as I have the length, I'm only ever going to load 11 (cloned) completed transactions into memory in one go. I don't know if the wallet lib loads everything in at once but on the app side i think this is solved by default 🤷‍♂

Yes, that length you get currently indicates how many elements are loaded into the data structure in memory. So to avoid that I will need to update the loading process to be paginated OR load them one at a time as you request a specific index.

**Comment by CjS77** [comment]
From an implementation point of view the FFI library devs need to talk carefully with the mobile devs on the most effective way to provide transaction entries across the FFI boundary. 

Mobile apps aren't really paginated like web apps are, but typically use a "Recycle view" that holds a moving window of data; a subset of which is displayed depending on the scroll position. 

There also needs to be some collaboration on how to handle updates of data since the mobile os's might expect things done in a certain way.

**Comment by talmahaj** [comment]
@philipr-za @CjS77 Thanks for walking me through this; it's completely clear now.

So, as long as both users are online, it's a non-issue because "it's a few ms between hitting the send button and having the change output available" so if need be we could just show a loading icon for a moment and then showing their actual available balance would match their expectations.

Do you think we could leave it as-is for now (that is, displaying the _currently_ available balance with the label always reading "Available Balance") and then we can handle the async transaction cases in a separate ticket? This'll give @Karim-tari and myself time to think about it without blocking you guys.

**Comment by CjS77** [comment]
When the receiver is not online though, how do you want to play it?

**Comment by philipr-za** [comment]
"it's a few ms between hitting the send button and having the change output available" is not strictly true. The change output will only be available to be used in another transaction after it has been mined which could take minutes.

**Comment by CjS77** [comment]
It doesn't even have to be together AFAIK. A basenode will accept the TX if it sees an input in the mempool, right @neonknight64 ?

**Comment by neonknight64** [comment]
A new second transaction that spends the change output of the first transaction can be constructed before the first transaction is mined. The mempool will keep the second transaction in the orphan pool until the change output becomes a valid UTXO at which point the second transaction becomes mineable. A future improvement could be to make them mineable at the same time.

**Comment by philipr-za** [comment]
@talmahaj I noted that in the ticket you only cater for transactions that are Pending and then completed. I wanted to clarify what these states mean in relation to the backend states that a transaction can have. 

Just for 100% clarity in this comment I wanted to reiterate the states a transaction will go through in the backend. So for a transaction that is being sent by a Sender to a Receiver the transaction will initially be sent and be waiting for the Receiver to reply, in the backend this is currently called a Pending Outbound Transaction. Once the Receiver replies the Sender can complete the transaction but it has not been broadcast to the mempool yet. Once it has been broadcast it still needs to be mined and then finally once it has been mined it is properly final in the blockchain (ideally there should be some confirmation time but lets leave that out of the equation for now). So in the backend a Sent transaction's states looks like this:
1. Send to Receiver -> Pending Outbound Transaction
2. Receiver has replied -> Completed transaction (Not broadcast to mempool)
3. Broadcast to Mempool -> Completed transaction (Broadcast)
4. Mined -> Completed Transaction (Mined)

Now I understand we want to abstract this in the front end so I wanted to ask how do we map the backend states to the frontend presentation? 

If we were just going with 2 frontend states (pending, complete) then the only real safe option is to have backend states 1,2 and 3 map to Pending and only after the transaction is Mined move to completed. However mining can take multiple minutes. So it feels like we should actually have three states on the frontend.
1. Waiting for response from other party
2. Pending in mempool
3. Completed (after mining)

What are your thoughts?

**Comment by talmahaj** [comment]
> When the receiver is not online though, how do you want to play it?

@CjS77 We don't know yet, and unfortunately, @Karim-tari has been pulled in so many directions that we haven't yet had any time to discuss and consider the different ways we could handle that case. I was planning on keeping it out of the scope of this ticket so as not to block dev, and creating a separate ticket for it that we can get back to (it's in the roadmap spreadsheet in the User Dashboard section).

> @talmahaj I noted that in the ticket you only cater for transactions that are Pending and then completed. I wanted to clarify what these states mean in relation to the backend states that a transaction can have.
> 
> Just for 100% clarity in this comment I wanted to reiterate the states a transaction will go through in the backend. So for a transaction that is being sent by a Sender to a Receiver the transaction will initially be sent and be waiting for the Receiver to reply, in the backend this is currently called a Pending Outbound Transaction. Once the Receiver replies the Sender can complete the transaction but it has not been broadcast to the mempool yet. Once it has been broadcast it still needs to be mined and then finally once it has been mined it is properly final in the blockchain (ideally there should be some confirmation time but lets leave that out of the equation for now). So in the backend a Sent transaction's states looks like this:
> 
> 1. Send to Receiver -> Pending Outbound Transaction
> 2. Receiver has replied -> Completed transaction (Not broadcast to mempool)
> 3. Broadcast to Mempool -> Completed transaction (Broadcast)
> 4. Mined -> Completed Transaction (Mined)
> 
> Now I understand we want to abstract this in the front end so I wanted to ask how do we map the backend states to the frontend presentation?
> 
> If we were just going with 2 frontend states (pending, complete) then the only real safe option is to have backend states 1,2 and 3 map to Pending and only after the transaction is Mined move to completed. However mining can take multiple minutes. So it feels like we should actually have three states on the frontend.
> 
> 1. Waiting for response from other party
> 2. Pending in mempool
> 3. Completed (after mining)
> 
> What are your thoughts?

@philipr-za Thanks for reiterating this, and for reminding me to mention it -- it's my oversight for not doing so. The states and sequence are completely clear to me -- Cayle has explained how the mining time distribution looks to me as well -- and just like the case where one user is offline, we just haven't yet had time to think through the UX around this in detail (and thus didn't intend for this ticket to handle more than binary complete vs. non-complete statuses for now. That is, as you mention, having states 1, 2, and 3 map to Pending for now and then revisiting when we get the chance). I agree that since mining can take multiple minutes, and could create uncertainty for the user, it's very likely that we'll go with the 3 FE states above as you suggest, and/or use some time-based feedback to assure the user of what's going on if mining takes more than a few minutes.

@Jasonvdb  and Kutsal FYI, we'll have tickets to handle the cases above coming separately, for now please map all the BE states as follows per Phil's note above:

Pending:
> 1. Send to Receiver -> Pending Outbound Transaction
> 2. Receiver has replied -> Completed transaction (Not broadcast to mempool)
> 3. Broadcast to Mempool -> Completed transaction (Broadcast)

Complete:
> 4. Mined -> Completed Transaction (Mined)

**Comment by talmahaj** [comment]
@Jasonvdb here are all the tweaks/additions I mentioned including the ones from yesterday. I've also added all of these to the original ticket body above so you and @kukabi can always refer to it as the source of truth for final specifications (rather than having to remember what's in the comments).

---

Updated Zeplin files for below can be found under Dec 3 Updates section: 
https://app.zeplin.io/project/5dbdf1a40bdfd3832e35c9c5/dashboard?seid=5de61c380c863977d92a7d0a

**Dec 1/2 tweaks**

- Sticky header
    - The expanded list UI should have a sticky header and a close button on the top left per https://zpl.io/aNn9wmv and https://zpl.io/V4dv0OX

- Scroll behavior for CTA
    - As soon as user starts scrolling back up in the tx list, no matter how far down they are, the Send Tari button should immediately appear again.

**Dec 3 - Finalized outstanding design tweaks**  

- List item font sizes bumped up for improved readability of both contact name and note (along with slight spacing tweaks to maintain proportions):  

- Edge case handling for large transactions (…not exactly a tweak, we just didn’t show these cases in prior mocks): https://zpl.io/2yEzvGK
    - Green pill length should be dynamic based on number of digits in amount
    - Truncation point where contact name and note display ellipses (…) should be dynamic based on number of digits in amount
    - Comma should appear when transaction in thousands of Tari
    - Two decimal places should always appear


**Dec 3 - Finalized missing animations**

- How transaction sections and transactions appear/disappear:
https://vimeo.com/showcase/6606770/video/376982133 pw: t@ri (ignore inconsistent typography, final styles are in Zeplin)
-   In sequence per the video:
    1. For every new section that appears — whether it is Pending Transactions OR a calendar day — the section heading should animate in at the same time as the transaction(s) within it
    2. When the Pending Transactions section appears, the animation next to the section header on the right should appear at the same time, and persist until there are no more pending transactions (and thus the entire section disappears out of view)
    3. If a section already exists, but it has a new list item, the section heading should first slide to accommodate that item before the new list item animates in
    4. When there are no more pending transactions, progress animation should become green checkmark before section disappears from view
    5. When Pending Transactions section disappears, now-complete transaction should animate into the calendar day section at the same time the whole list slides up

- Page load animations
Not a tweak, just confirming that whenever the transaction list loads, its elements should still animate in per 00:12/13 in the first prototype we did: https://vimeo.com/369345180 

- Dynamic balance changes
Not a tweak, just confirming that whenever the available balance changes, any changed digits should still animate in per 01:25 in the first prototype we did: https://vimeo.com/369345180


Dec 3 - Other

- Dev to confirm the transactions move to the correct sections if I switch my phone’s time zone such that a transaction is now on a prior or future day

**Comment by talmahaj** [comment]
@Karim-tari When transaction list is in full screen view, if user is at the top of the list already and swipes down, should it collapse back into its default state? Seems to me that it should work in reverse too... 

cc @Jasonvdb @kukabi

**Comment by Karim-tari** [comment]
Yeah agreed. I think it should.

**Comment by Jasonvdb** [comment]
@talmahaj that would conflict with the pull to refresh gesture then. So if we still want pull to refresh it would need to be when the transaction list was already swiped down. Would that be fine?

**Comment by kukabi** [comment]
I agree with @Jasonvdb with pull to refresh - it can only be available in the swiped down mode.

**Comment by kukabi** [comment]
Are we still going to have the search bar as in the video? @talmahaj @Jasonvdb

**Comment by Jasonvdb** [comment]
@talmahaj I'm going ahead with this for the blank slate https://zpl.io/aN5eEve 
I know the rest of the screen's design is outdated but I don't see an updated one so using the image and text. Let me know if that's wrong
